text/0000-tracked-properties.md

```diff
@@ -0,0 +1,1049 @@
+  and not having to specify dependent keys.
+
+  ```js
+  import { tracked, computed } from '@ember/object';
+
+  class Person {
+    firstName;
+    lastName;
+
+    @tracked middleName;
+
+    @computed('firstName', 'lastName')
+    get fullName() {
+      return `${this.firstName} ${this.middleName} ${this.lastName}`;
+    }
+  }
+
+  let person = new Person();
+
+  set(person, 'firstName', 'Tom');
+  set(person, 'lastName', 'Dale');
+
+  person.middleName = 'Tomster';
+
+  person.fullName; // 'Tom Tomster Dale'
+  ```
+
+#### `get` and `set`
+
+It is common in the legacy model to set and consume plain object properties
+which are not computed properties, or in any other way special. Ember's `get`
+and `set` functions historically allowed this by giving us the ability to
+intercept all property changes and watch for mutations.
+
+This presents a problem for tracked properties, particularly because of the
+recent change in Ember to enable native Javascript getters to replace `get`.
+This change means that we have no way to intercept `get`, and consequently no
+way for tracked properties to know whether or not a plain property will later
+be updated with `set`.
+
+To demonstrate this case, consider the following service and component:
+
+```js
+const Config = Service.extend({
+  polling: {
+    shouldPoll: false,
+    pollInterval: -1,
+  },
+
+  init() {
+    this._super(...arguments);
+
+    fetch('config/api/url')
+      .then(r => r.json())
+      .then(polling => set(this, 'polling', polling));
+  },
+});
+```
+```js
+class SomeComponent extends Component {
+  @service config;
+
+  @tracked
+  get pollInterval() {
+    let { shouldPoll, pollInterval } = this.config.polling;
+
+    return shouldPoll ? pollInterval : -1;
+  }
+}
+```
+
+Let's walk through the flow here:
+
+1. The `SomeComponent` component is rendered for the first time, instantiating
+   the `Config` service (assuming this the first time it has ever been
+   accessed). The service's init hook kicks off an async request to get the
+   configuration from a remote URl.
+2. The tracked `pollInterval` property first accesses the service injection,
+   which is a computed property. The property is detected and added to the
+   tracked stack.
+3. We then access the plain, undecorated `polling` object. Because it is
+   is not tracked and not a computed property, we add the immutable setter which
+   will throw an error.
+4. Sometime later, the async request returns with the configuration object. We
+   attempt to set it on the service, and trigger the immutable setter assertion.
+
+In order to prevent this from happening, user's will have to use `get` when
+accessing any values which may be set with `set`, and are not computed
+properties.
+
+```js
+class SomeComponent extends Component {
+  @service config;
+
+  @tracked
+  get pollInterval() {
+    let shouldPoll = get(this, 'config.polling.shouldPoll');
```

Comment:
I really don't like this

Consider the follow:

- Service property goes from from "static" to "computed", now all consumers need to update.
- To understand if you need to use `get` you need to look at the implementation. This is not obvious for addon code. Will fuel "always you `get`". This is implementation leakage.
- The tracked property maybe pulled on lazily e.g. untested conditionals leading to way more bugs in production.

I'm wondering if the behavior of `@service` can actually do something like lazily install `tracked`.

Reply:
As far as I've thought it through, I believe it is not possible in a performant way. The issue is that without native proxies, we need to install property getters for every link in between our tracked property and whichever services it accesses.

So, let's break this down:

```js
const Config = Service.extend({
  polling: {}
});
```
```js
class SomeComponent extends Component {
  @service config;

  @tracked
  get pollInterval() {
    let config = this.config;
    let polling = config.polling;

    let shouldPoll = polling.shouldPoll;
    let pollInterval = polling.pollInterval;

    return shouldPoll ? pollInterval : -1;
  }
}
```

1. First, we access `this.config`. We decorated this getter, it's an injection, so we can do whatever we want here. But what do we do? We have no idea what the user will access on this service. We could either:
  A. Attempt to introspect on the body of `pollInterval`, the tracked function context. The only way we could do this is by using `toString()`, which seems like it would be very brittle. Alternatively, we may be able to introspect at build time using Babel, but this also seems problematic.
  B. Add a native getter to _every_ known property on the service. This seems very non-performant, since it's doing much more work than we would want to do, but it would get the job done.
2. Assuming we can instrument the `polling` property properly, when we access it we now need to instrument the object returned by the property. In this version of the example, I purposefully removed the default configuration values because it's not uncommon to see Ember users do this, and in the pure `get`/`set` world, it _just works_. But now, we can't even blanket apply getters. Introspection on the getter body is our only option.

There is a third option here, but it requires some constraints on the `get`/`set` side. We could entangle the service's object tag when the service is retrieved from the injection with the tracked property, such that _all_ changes to the service will invalidate the property. However, for this to work, we need to always `set` from the context of the service. This would break, because `set` would have no way of knowing that it should dirty the service:

```js
const Config = Service.extend({
  polling: {
    shouldPoll: false,
    pollInterval: -1,
  },

  init() {
    this._super(...arguments);

    fetch('config/api/url')
      .then(r => r.json())
      .then(({ shouldPoll, pollInterval }) => {
        set(this.polling, 'shouldPoll', shouldPoll);
        set(this.polling, 'pollInterval', pollInterval);
      });
  },
});
```

Reply:
Even with native proxies, we may not be able to achieve full safety. You can see the recent diff that explains this here, but it comes down to a decision by TC39 to prioritize security over expressivity with native proxies: https://github.com/tomdale/rfcs/pull/6